Remove UDP support from the network tests

thsfs opened the pull request:
Migrates or remove network tests that use UDP.

- Adds a helper function to add a system node disconnected from its network.
- Organized the client/server of the tcp_connection test to be in distinct classes.
- Removed tests that are exclusive to UDP, otherwise migrated them to TCP or inproc.

Fixes issue: https://github.com/nanocurrency/nano-node/issues/3846

Commit:
Organizes the tcp_connection test into client/server classes

pwojcikdev commented on nano/core_test/network.cpp:
```diff
@@ -434,96 +397,44 @@ TEST (network, send_insufficient_work)
 	ASSERT_EQ (2, node2.stats.count (nano::stat::type::error, nano::stat::detail::insufficient_work));
 }
 
-TEST (receivable_processor, confirm_insufficient_pos)
```
Why remove `confirm_sufficient_pos` & `confirm_insufficient_pos` tests?

pwojcikdev replied:
Those test were seemingly missing ASSERTS but perhaps they were testing internal debug asserts inside message processing pipeline?

dsiganos replied:
I think someone just forgot to place the final asserts.
I think pos stands for Proof Of Stake in this case.
The firsts tires to send a vote with no weight and the seconds sends a vote with 100% weight.

dsiganos replied:
I think someone just forgot to place the final asserts.
I think pos stands for Proof Of Stake in this case.
The firsts tires to send a vote with no weight and the seconds sends a vote with 100% weight.

pwojcikdev commented on nano/core_test/network.cpp:
```diff
@@ -977,35 +878,6 @@ TEST (tcp_listener, tcp_listener_timeout_node_id_handshake)
 	}
 }
 
-TEST (network, replace_port)
```
Can we port this test to TCP? It should ensure when we connect a new realtime tcp channel to the node from different port the old one is replaced. This is important behavior.

pwojcikdev replied:
As @thsfs correctly pointed out, for TCP connections we replace the whole connection, not just the port. This is not needed and can be removed.

dsiganos commented on nano/core_test/network.cpp:
```diff
@@ -1,57 +1,104 @@
-	connector.async_connect (boost::asio::ip::tcp::endpoint (boost::asio::ip::address_v4::loopback (), acceptor.local_endpoint ().port ()),
-	[&done2, &message2] (boost::system::error_code const & ec_a) {
-		if (ec_a)
+	// Organizes the code used for a connection into distinguished classes (base class/client/server)
```
Class simple_socket looks very generic and probably useful for other tests too.
I think it should be moved into its own file in a test library.

thsfs replied:
I did this before, but as this is used only by this test decided to have it only here. Prefer to do the move only if this is shared with other tests. Not a change for now.

dsiganos commented on nano/core_test/network.cpp:
```diff
@@ -1,57 +1,104 @@
 
 using namespace std::chrono_literals;
 
 TEST (network, tcp_connection)
```
Add a small comment describing what this test does

dsiganos commented on nano/core_test/network.cpp:
```diff
@@ -1,57 +1,104 @@
+			accept ();
+		}
+
+		void accept ()
```
Usually accept can be called many times whereas as this accept implementation expects to be called just once. It is a bit confusing to call this step accept. Maybe call it accept_once_only to remove the confusion? Probably this is not a major issue.

dsiganos replied:
I just noticed that it also does a receive too. So it does:
* init
* listen
* accept
* receive

thsfs replied:
That is another reason I didn't promote this code to the test library, it just organizes better the code we had before in this test.

dsiganos commented on nano/core_test/network.cpp:
```diff
@@ -74,48 +121,6 @@ TEST (network, construction_without_specified_port)
 	EXPECT_EQ (port, node->bootstrap.endpoint ().port ());
 }
 
-TEST (network, self_discard)
```
Why was this test removed? The concept of self-connecting exists in TCP too.

thsfs replied:
This seems to be purely UDP. As we have not been testing TCP for this, I think the best approach is to create an issue for us to create a test for it as a further step. Thinking about to do the same with the other tests I removed. The criteria used to remove them as that they were either not testing anything or that seemed to be UDP exclusive.

dsiganos replied:
Let's take the opposite approach, we do not remove tests unless we are sure we do not need them.
so let's reinstate this test as it was, using UDP, and remove or convert in a different PR.

dsiganos commented on nano/core_test/network.cpp:
```diff
@@ -1056,27 +928,23 @@ namespace transport
-	ASSERT_EQ (1, node1.stats.count (nano::stat::type::filter, nano::stat::detail::duplicate_publish));
 	tcp_channel->send (publish);
-	ASSERT_TIMELY (2s, node1.stats.count (nano::stat::type::filter, nano::stat::detail::duplicate_publish) == 2);
+	ASSERT_TIMELY (2s, node1.stats.count (nano::stat::type::filter, nano::stat::detail::duplicate_publish) == 0);
```
2 seconds is too little on a very slow computer. We have agreed in a meeting to wait at least 5 seconds for these events.

dsiganos commented:
Thiago, it appears that most objections are about the rmoeval of tests.
So let's reinstate them as they were (using UDP) and get traction on this Pr, which on the whole is good and almost ready to merge.

Commit:
Remove exclusive UDP tests or migrate (core_test/network.cpp)

Commit:
Remove network.replace_port test

The port replacement doesn't happen for TCP, only for UDP. In TCP the
whole channel is replaced.
